polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp

```diff
@@ -703,13 +710,218 @@ void AffineParallelGuardBuilder::getConstraints(
+              builder.getTypeArrayAttr({accessor.getType(), id.getType()}));
+    attrs.set(mlir::sycl::SYCLDialect::getFunctionNameAttrName(),
+              FlatSymbolRefAttr::get(builder.getStringAttr("operator[]")));
+    attrs.set(mlir::sycl::SYCLDialect::getTypeNameAttrName(),
+              FlatSymbolRefAttr::get(builder.getStringAttr("accessor")));
+    return builder.create<sycl::SYCLAccessorSubscriptOp>(
+        loc, MT, ValueRange({accessor, id}), attrs);
+  }
+
+  static sycl::SYCLAccessorSubscriptOp
+  getSYCLAccessorBegin(TypedValue<MemRefType> accessor, OpBuilder builder,
+                       Location loc) {
+    const auto accTy =
+        cast<sycl::AccessorType>(accessor.getType().getElementType());
+    const auto idTy = cast<sycl::IDType>(
+        cast<sycl::AccessorImplDeviceType>(accTy.getBody()[0]).getBody()[0]);
+    auto id = builder.create<memref::AllocaOp>(loc, MemRefType::get(1, idTy));
+    const Value zero = builder.create<arith::ConstantIntOp>(loc, 0, 64);
+    const Value zeroIndex = builder.create<arith::ConstantIndexOp>(loc, 0);
+    for (unsigned i = 0; i < accTy.getDimension(); ++i) {
+      sycl::SYCLIDGetOp idGetOp = createSYCLIDGetOp(id, i, builder, loc);
+      builder.create<memref::StoreOp>(loc, zero, idGetOp, zeroIndex);
+    }
+    return createSYCLAccessorSubscriptOp(accessor, id, builder, loc);
+  }
+
+  static polygeist::SubIndexOp
+  getSYCLAccessorEnd(TypedValue<MemRefType> accessor, OpBuilder builder,
+                     Location loc) {
+    const auto accTy =
+        cast<sycl::AccessorType>(accessor.getType().getElementType());
+    sycl::SYCLAccessorGetRangeOp getRangeOp =
+        createSYCLAccessorGetRangeOp(accessor, builder, loc);
+    auto range = builder.create<memref::AllocaOp>(
+        loc, MemRefType::get(1, getRangeOp.getType()));
+    const Value zeroIndex = builder.create<arith::ConstantIndexOp>(loc, 0);
+    builder.create<memref::StoreOp>(loc, getRangeOp, range, zeroIndex);
+    const auto idTy = cast<sycl::IDType>(
+        cast<sycl::AccessorImplDeviceType>(accTy.getBody()[0]).getBody()[0]);
+    auto id = builder.create<memref::AllocaOp>(loc, MemRefType::get(1, idTy));
+    const Value one = builder.create<arith::ConstantIntOp>(loc, 1, 64);
+    for (unsigned i = 0; i < accTy.getDimension(); ++i) {
+      sycl::SYCLIDGetOp idGetOp = createSYCLIDGetOp(id, i, builder, loc);
+      sycl::SYCLRangeGetOp rangeGetOp =
+          createSYCLRangeGetOp(range, i, builder, loc);
+      auto index = builder.create<arith::SubIOp>(loc, rangeGetOp, one);
+      builder.create<memref::StoreOp>(loc, index, idGetOp, zeroIndex);
+    }
+    auto accSub = createSYCLAccessorSubscriptOp(accessor, id, builder, loc);
+    const Value oneIndex = builder.create<arith::ConstantIndexOp>(loc, 1);
+    return builder.create<polygeist::SubIndexOp>(loc, accSub.getType(), accSub,
+                                                 oneIndex);
```

Comment:
Wouldn't the end be `accSub`?

Reply:
not really, let say accessor A has get_range = {3, 3}, the end pointer should be &A[{2, 2}] + 1, and not &A[{3, 3}], as that would be out of bound, and &A[{2, 2}] would be the beginning of the last element.

Reply:
True. Can we not use `polygeist::SubIndexOp`, but building the input ID taking that into account?

Reply:
Subtract 1 from the last dimension of `get_range()` when building the offset ID

Reply:
oic! do you mean plus 1 from the last dimension?

Reply:
Yes, that's what I meant!